tsl/src/continuous_aggs/create.c

```diff
@@ -1348,40 +1348,50 @@ replace_having_qual_mutator(Node *node, cagg_havingcxt *cxt)
 {
 	if (node == NULL)
 		return NULL;
-	if (equal(node, cxt->old->expr))
+	ListCell *lc, *lc2;
+	List *origtlist = cxt->orig_tlist;
+	List *modtlist = cxt->mod_tlist;
+	forboth (lc, origtlist, lc2, modtlist)
 	{
-		cxt->found = true;
-		return (Node *) cxt->new->expr;
+		TargetEntry *te = (TargetEntry *) lfirst(lc);
+		TargetEntry *modte = (TargetEntry *) lfirst(lc2);
+		if (equal(node, te->expr))
+		{
+			cxt->found = true;
+			return (Node *) modte->expr;
+		}
 	}
 	return expression_tree_mutator(node, replace_having_qual_mutator, cxt);
 }
 
-/* modify the havingqual and replace exprs that already occur in targetlist
- * with entries from new target list
+/* modify the havingqual and replace exprs (in havingqual) that already occur
+ * in targetlist with entries from new target list
+ * Arguments:
+ *   origquery : Query whose havingqual will be modified.
+ *   newtlist : fixed up targetlist for origquery (the origquery's tlist is mapped
+ *              to columns from materialized hypertable). There is a 1-1 mapping
+ *              between origquery->targetList and newtlist
  * RETURNS: havingQual
+ * Example:
+ * SELECT x, count(x) , time_bucket(...)
+ * FROM ...
+ * HAVING count(x) > 10 and sum(y) > 11 --->havingQual
+ *
+ * count(x) : already appears in the targetlist of the query and should be
+ * replaced by the corresponding entry from modtlist. We have to compare count(x)
+ * with all the entries in the tlist so that we do not match subexprs when the
+ * complete expr (e.g do not match x instead of count(x) ) is in the tlist.
+ * (issue 2655)
```

Comment:
This is not particularly easy to follow. If you add the `mod_tlist` and `orig_tlist` for the example above it will be easier to follow how the replacement works.

Also, do you have a test case for the situation you describe above, that is, a test case that would break if you do not compare `count(x)` with all the entries in the tlist.

Reply:
I was wondering about the specific corner-case where you have both the subexpression and the "top-expression" in the tlist.

Reply:
Both search_query and count(search_query) appear in the tlist (in the query)
HAVING has count(search-query)
```
CREATE MATERIALIZED VIEW search_query_count_1m WITH (timescaledb.continuous) 
AS
 SELECT  search_query,count(search_query) as count, 
         time_bucket(INTERVAL '1 minute', time) AS bucket
 FROM raw_data
 WHERE search_query is not null AND LENGTH(TRIM(both from search_query))>0
 GROUP BY search_query, bucket 
HAVING count(search_query) > 3 OR sum(cnt) > 1;
```

Reply:
Thank you, then it is covered.

Reply:
Could you please show the tlist and the modlist in the example as well, so that it is clear to future developers?

Reply:
added more comments